Filter primitive subregion with relative values should properly render referenced elements

nullhook opened the pull request:
#### 4c23e972c221
<pre>
Filter primitive subregion with relative values should properly render referenced elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=298540">https://bugs.webkit.org/show_bug.cgi?id=298540</a>
<a href="https://rdar.apple.com/154363710">rdar://154363710</a>

Reviewed by NOBODY (OOPS!).

SVG filters can be defined independently and referenced via CSS
or another SVG. In this scenario, when a filter primitive uses a
percentage-based subregion, the current implementation fails to render
the referenced element due to empty viewport information.

To solve this, we fallback to object bounding box behavior and use the
given viewport size.

* LayoutTests/svg/filters/svg-filter-external-references-expected.html: Added.
* LayoutTests/svg/filters/svg-filter-external-references.html: Added.
* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::dimensionForLengthMode):
(WebCore::SVGLengthContext::resolveRectangle):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4c23e972c22175a7a963ad49bba387207d8d3831

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121120 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40816 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31474 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127539 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73201 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/99a29866-7521-482f-a885-065af8098536) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122996 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41518 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49395 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92006 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61211 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/75f39298-7756-4192-b3d1-0bffbde3f2fa) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124072 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33151 "Found 1 new test failure: svg/filters/svg-filter-external-references.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108568 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72689 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c7bbdf68-9f96-44e7-9ed6-d493a7750ccf) 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32175 "Exiting early after 10 failures. 10 tests run. 1 failures") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26671 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71131 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102656 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26850 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130392 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48047 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36518 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100614 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48415 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104735 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100517 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45918 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23968 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44741 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47905 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53618 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47376 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50723 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49060 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a5cac5ea1c7d3cede9f7c3afc8d066935e334901. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50452

nullhook commented:
there seems to be clipping issues:


1. when SVG filter is referenced via CSS regardless of the units.

<img width="202" height="203" alt="image" src="https://github.com/user-attachments/assets/343278ff-4b44-40da-9442-b9a48882fb0e" />

on initial load you'll see the image hard clipped

<img width="200" height="202" alt="image" src="https://github.com/user-attachments/assets/e3263787-7901-4634-b2f5-cb29fc71cb96" />

after user invalidates (focus/zoom) 

2. when using CSS filter functions ex. blur(...)

<img width="201" height="203" alt="image" src="https://github.com/user-attachments/assets/e97c8c31-23ea-4842-9401-df68b58fc0bc" />

on initial load

<img width="202" height="204" alt="image" src="https://github.com/user-attachments/assets/a7764915-ccbf-4986-9446-8315c329a7b2" />

after invalidation

nullhook commented on LayoutTests/svg/filters/svg-filter-references.html:
```diff
@@ -0,0 +1,39 @@
+<body>
+    <svg width="0" height="0" style="position: absolute;">
+        <defs>
+            <filter id="blurFilter">
```
shouldn't stdDeviation change in this config? ideally to 0.05.

nullhook replied:
updated to `objectBoundingBox`

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/62654ee21559d0195f1fd8b1aecdae2ac9771974. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50452

nullhook committed:
Filter primitive subregion with relative values should properly render referenced elements
https://bugs.webkit.org/show_bug.cgi?id=298540
rdar://154363710

Reviewed by NOBODY (OOPS!).

SVG filters can be defined independently and referenced via CSS
or another SVG. In this scenario, when a filter primitive uses a
percentage-based subregion, the current implementation fails to render
the referenced element due to empty viewport information.

To solve this, we fallback to object bounding box behavior and use the
given viewport size.

* LayoutTests/svg/filters/svg-filter-external-references-expected.html: Added.
* LayoutTests/svg/filters/svg-filter-external-references.html: Added.
* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::dimensionForLengthMode):
(WebCore::SVGLengthContext::resolveRectangle):

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/731b2f22b3403d3ea446dbbefe1ecf73ca618bd8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50452

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e6088077bf166377f6661a8407ac09636fa6ca74. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50452

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4c23e972c22175a7a963ad49bba387207d8d3831. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50452
